File: amethyst/Domain/ActivityData.cs
namespace amethyst.Domain;

public record ActivityData(ClientActivity Activity, string? GameId, string LanguageCode);
public record UnknownActivity() : ActivityData(ClientActivity.Unknown, null, "en");
public record ScoreboardActivity(string? GameId, string LanguageCode, bool UseSidebars, bool UseNameBackgrounds) : ActivityData(ClientActivity.Scoreboard, GameId, LanguageCode);
public record StreamOverlayActivity(string? GameId, string LanguageCode, double Scale, bool UseBackground, string BackgroundColor) : ActivityData(ClientActivity.StreamOverlay, GameId, LanguageCode);
public record PenaltyWhiteboardActivity(string? GameId, string LanguageCode) : ActivityData(ClientActivity.PenaltyWhiteboard, GameId, LanguageCode);
public record ScoreboardOperatorActivity(string? GameId, string LanguageCode) : ActivityData(ClientActivity.ScoreboardOperator, GameId, LanguageCode);
public record PenaltyLineupControlActivity(string? GameId, string LanguageCode) : ActivityData(ClientActivity.PenaltyLineupControl, GameId, LanguageCode);
public record PenaltyControlActivity(string? GameId, string LanguageCode) : ActivityData(ClientActivity.PenaltyControl, GameId, LanguageCode);
public record LineupControlActivity(string? GameId, string LanguageCode) : ActivityData(ClientActivity.LineupControl, GameId, LanguageCode);
public record BoxTimingActivity(string? GameId, string LanguageCode) : ActivityData(ClientActivity.BoxTiming, GameId, LanguageCode);
public record OtherActivity(string? GameId, string LanguageCode) : ActivityData(ClientActivity.Other, GameId, LanguageCode);

public enum ClientActivity
{
    Unknown,
    Scoreboard,
    StreamOverlay,
    PenaltyWhiteboard,
    ScoreboardOperator,
    PenaltyLineupControl,
    PenaltyControl,
    LineupControl,
    BoxTiming,
    Other,
}

File: jamster.ui/packages/shared/src/types/ActivityData.ts
import { ClientActivity } from "./ClientActivity";

export type ActivityData = {
    gameId: string | null;
    languageCode: string;
} & (
    UnknownActivity 
    | ScoreboardActivity 
    | StreamOverlayActivity 
    | PenaltyWhiteboardActivity 
    | ScoreboardOperatorActivity 
    | PenaltyLineupControlActivity
    | PenaltyControlActivity
    | LineupControlActivity
    | BoxTimingActivity
    | OtherActivity
)

export type UnknownActivity = {
    activity: ClientActivity.Unknown;
}

export type ScoreboardActivity = {
    activity: ClientActivity.Scoreboard;
    useSidebars: boolean;
    useNameBackgrounds: boolean;
}

export type StreamOverlayActivity = {
    activity: ClientActivity.StreamOverlay;
    scale: number;
    useBackground: boolean;
    backgroundColor: string;
}

export type PenaltyWhiteboardActivity = {
    activity: ClientActivity.PenaltyWhiteboard;
}

export type ScoreboardOperatorActivity = {
    activity: ClientActivity.ScoreboardOperator;
}

export type PenaltyLineupControlActivity = {
    activity: ClientActivity.PenaltyLineupControl;
}

export type PenaltyControlActivity = {
    activity: ClientActivity.PenaltyControl;
}

export type LineupControlActivity = {
    activity: ClientActivity.LineupControl;
}

export type BoxTimingActivity = {
    activity: ClientActivity.BoxTiming;
}

export type OtherActivity = {
    activity: ClientActivity.Other;
}

